mayastor-test/e2e/node_disconnect/node_disconnect_setup/node_disconnect_setup_test.go

```diff
@@ -0,0 +1,114 @@
+package node_disconnect_setup_test
+
+import (
+	"e2e-basic/common"
+	"fmt"
+	"sort"
+
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+const mayastor_regexp = "^mayastor-.....$"
+const moac_regexp = "^moac-..........-.....$"
+const namespace = "mayastor"
+const timeoutSeconds = 100
+
+// Set up for disconnection tests. Ensure moac is on the refuge node but
+// no mayastor instances are
+func disconnectSetupTest() {
+	// ensure we are using 2 replicas
+	common.MkStorageClass("mayastor-iscsi-2", 2, "iscsi", "io.openebs.csi-mayastor")
+	common.MkStorageClass("mayastor-nvmf-2", 2, "nvmf", "io.openebs.csi-mayastor")
+
+	nodeList, err := common.GetNodeLocs()
+	Expect(err).ToNot(HaveOccurred())
+	Expect(len(nodeList) >= 3)
+
+	// sort the nodes - that also means k8s-1 is the refuge on local clusters
+	sort.Slice(nodeList, func(i, j int) bool { return nodeList[i].NodeName < nodeList[j].NodeName })
+	refugeIndex := 0
+
+	// Select one node to be the refuge, remove the engine=mayastor label so mayastor does not run there
+	refugeNode := ""
+	for i, node := range nodeList {
+		if i == refugeIndex {
+			refugeNode = node.NodeName
+			common.UnlabelNode(refugeNode, "openebs.io/engine")
+			common.LabelNode(refugeNode, "openebs.io/podrefuge=true")
+		}
+	}
+	Expect(refugeNode != "")
+
+	// reduce the number of moac instances to be zero
+	// this seems to be needed to guarantee that moac moves to the refuge node
+	var repl int32 = 0
+	common.SetDeploymentReplication("moac", namespace, &repl)
+
+	// wait for moac to disappear from the cluster
+	for _, node := range nodeList {
+		if node.NodeName != refugeNode {
```

Comment:
Is it possible that moac is already running on the refuge node? In which case we will bring up a moac instance before the old one is gone. Even if this is the case, maybe it doesn't matter?

Reply:
The number of replicas has been set to zero on line 51, so ultimately, there should be no moacs running before the loop is exited (even if there was one on another node - which shouldn't happen).

Reply:
The code just changes the number of replicas to zero. If it was already on the refuge node the loop will exit immediately. Then the number of replicas is set to 1, but moac will remain on that node because the node-affinity forces it to be there.

Reply:
Pushed change to  move moac only if not already on the refuge node.